File: 2025/day5/p2.py
def main():
    fin = open("input.txt", "r")

    one = []
    two = []
    innit = False

    for line in fin:
        line = line.rstrip('\n')
        if line == "":
            innit = True
            continue
        if not innit:
            one.append(line)
        else:
            two.append(line)

    ranges = []
    for s in one:
        L = 0
        R = 0
        dash = False
        for c in s:
            if c == '-':
                dash = True
                continue
            if not dash:
                L = L * 10 + (ord(c) - ord('0'))
            else:
                R = R * 10 + (ord(c) - ord('0'))
        ranges.append((L, R))

    ranges.sort()
    merged = []

    for L, R in ranges:
        if not merged:
            merged.append([L, R])
            continue

        lastL, lastR = merged[-1]
        if L <= lastR + 1:
            merged[-1][1] = max(lastR, R)
        else:
            merged.append([L, R])

    total_fresh = 0
    for L, R in merged:
        total_fresh += (R - L + 1)

    print(total_fresh)


if __name__ == "__main__":
    main()

File: 2025/day5/p2.cpp
#include <bits/stdc++.h>
using namespace std;
#define fast_io ios::sync_with_stdio(false);cin.tie(nullptr);

int main(){
    fast_io;

    ifstream fin("input.txt");

    vector<string> one;
    vector<string> two;
    string line;
    bool innit = false;
    set<long long> sex;

    while (getline(fin, line)){
        if (line == ""){
            innit = true;
            continue;
        }
        if (!innit){
            one.push_back(line);
        } else {
            two.push_back(line);
        }
    }

    vector<pair<long long,long long>> ranges;

    long long totalWork = 0;
    for (auto &s : one) {
        long long L = 0, R = 0;
        bool dash = false;
        for (char c : s) {
            if (c == '-') { dash = true; continue; }
            if (!dash) L = L * 10 + (c - '0');
            else       R = R * 10 + (c - '0');
        }
        ranges.push_back({L, R});
        totalWork += (R - L + 1);
    }

    long long done = 0;

    for (auto &p : ranges) {
        long long L = p.first;
        long long R = p.second;

        for (long long z = L; z <= R; z++) {
            sex.insert(z);
            done++;

            if (done % 1000000 == 0) {
                double pct = (double)done * 100.0 / totalWork;
                cout << "\rProgress: " 
                     << fixed << setprecision(2) 
                     << pct << "%   " << flush;
            }
        }
    }

    cout << "\n" << sex.size() << "\n";
    return 0;
}

